taskchampion/docs/src/encryption.md

```diff
@@ -0,0 +1,38 @@
+# Encryption
+
+The client configuration includes an encryption secret of arbitrary length.
+This section describes how that information is used to encrypt and decrypt data sent to the server (versions and snapshots).
+
+Encryption is not used for local (on-disk) sync, but is used for all cases where data is sent from the local host.
+
+## Key Derivation
+
+The client derives the 32-byte encryption key from the configured encryption secret using PBKDF2 with HMAC-SHA256 and 100,000 iterations.
+The salt value depends on the implemenation of the protocol, as described in subsequent chapters.
+
```

Comment:
Yeah, I think using the same salt for all users isn't good -- but I don't think there's any value to use? If we use the bucket name, then users can't change buckets. I'm open to ideas! @dbr I think you had some input when we decided to use PBKDF? What do you think?

Reply:
IIRC, the reason to use a KDF was basically to get good key entropy from the uneven entropy of human-invented passwords. So for `hunter2` instead of 68756e74657232000000000000000000 we'd get a decent bit pattern. Within a KDF, salt serves the purpose of resisting rainbow-table attacks and ensuring that even if two users choose the same password, they will use different keys.

For a rainbow table, the use of PBKDF in particular makes generation of a rainbow table for the fixed salt value `"taskchampion"` infeasible.

Differentiating users with the same encryption password is a small benefit from the cost of requiring some additional unique-to-the-user value. If we make the user supply that additional value, then there's still a likelihood that they will all pick the same value, e.g., from a sample config. And if we tie the value to something else in the config, such as the bucket name, then that cannot be changed later.

That's my understanding of things, but I'm happy to learn that I'm wrong :)

Reply:
> IIRC, the reason to use a KDF was basically to get good key entropy from the uneven entropy of human-invented passwords. 

We're basically on the same page here but I think it might be more accurate to say that KDF takes the entropy from a human-invented password and reformats it to be cryptographically random. I'm not sure a deterministic function can actually add entropy.

> Within a KDF, salt serves the purpose of resisting rainbow-table attacks and ensuring that even if two users choose the same password, they will use different keys.

> For a rainbow table, the use of PBKDF in particular makes generation of a rainbow table for the fixed salt value "taskchampion" infeasible.

I don't follow this. If the point of the salt is to prevent rainbow table generation doesn't reuse of the salt mean that a rainbow table *could* be made for a fixed salt value?

> Differentiating users with the same encryption password is a small benefit from the cost of requiring some additional unique-to-the-user value. If we make the user supply that additional value, then there's still a likelihood that they will all pick the same value, e.g., from a sample config. And if we tie the value to something else in the config, such as the bucket name, then that cannot be changed later.

This I do agree with and I think it might even be an understatement. From my understanding the whole premise of salt is that it is not stored by the same entity as the (decrypted) secret. In our case it seems like a salt would effectively be extending the secret across two different variables in the same configuration file, which is needlessly fragile.

Reply:
> .. it might be more accurate ..

Thanks, that's a much more accurate description!

> I don't follow this. If the point of the salt is to prevent rainbow table generation doesn't reuse of the salt mean that a rainbow table could be made for a fixed salt value?

I think PBKDF's 100ms+ per derivation makes generation of a rainbow table for one particular application infeasible. But I suppose a state actor could make one.

> From my understanding the whole premise of salt is that it is not stored by the same entity as the (decrypted) secret.

I'm not sure what this means. My understanding is that salts are usually used to ensure that two identical inputs to the same one-way function do not produce the same output, and the salt is included in the clear alongside the result of that one-way function. We're using PBKDF as a one-way function, but we're not storing the result of that function anywhere, so I'm not sure this understanding is helpful in this case. Maybe that means salts aren't helpful, either?

> .. it seems like a salt would effectively be extending the secret across two different variables in the same configuration file, which is needlessly fragile

Rephrasing to check that I understand: requiring that a user get _two_ configuration variables right in order to decrypt their data is worse for the user than requiring that they get _one_ configuration variable right.

Reply:
> I think PBKDF's 100ms+ per derivation makes generation of a rainbow table for one particular application infeasible. 

Not entirely related but I noticed we're using 100,000 iterations and [wikipedia](https://en.wikipedia.org/wiki/PBKDF2) says "In 2023, [OWASP](https://en.wikipedia.org/wiki/OWASP) recommended to use 600,000 iterations for PBKDF2-HMAC-SHA256..." The lack of unique salts might suggest we need even more iterations.

> But I suppose a state actor could make one.

Do nation state actors have more computational resources than major cloud providers? Is the cloud provider itself not the primary threat?

> My understanding is that salts are usually used to ensure that two identical inputs to the same one-way function do not produce the same output, and the salt is included in the clear alongside the result of that one-way function.

My recollection is that password-based kdf's are designed around server-side password storage. The server stores (i) the salt in the clear and (ii) the hash of password+salt. If an unsalted password database were compromised, an attacker could use a pre-computed rainbow table mapping hash to password to derive the password from the hash. If the password database was salted entirely with one salt, the attacker would have to construct a new rainbow table to attack this database. If every password is uniquely salted, the attacker would have to construct a new rainbow table for each password. (The whole premise of this system is that the password is the valuable thing, not the hashed password used to authenticate the user. That's because the web service can't be authenticated with the password hash, the attacker needs the password itself.) The point I was trying to make is that the server operator never has the cleartext password and the client doesn't have to store the salt. If the same entity is relied upon to store both values then there is no point in separating them.

So in our case, let's say a client's password is 'abc' and it is salted with '123'. I don't think that is any harder to crack than a password of 'abc123'. Even if this example isn't rigorously true, I think the principle stands that for any entropy offered by the client in the form of a salt it would be equally effective to add that entropy to the secret itself.

> Rephrasing to check that I understand: requiring that a user get two configuration variables right in order to decrypt their data is worse for the user than requiring that they get one configuration variable right.

Yes, a second variable is a usability issue but I think it's also a misrepresentation to call it a salt because it doesn't perform the function of a salt.

To summarize, my current position is that reusing salts does have *some* value insofar as a rainbow table would have to be constructed specifically for taskwarrior. But really, the recommendation should be strong generated random passwords. I'm doubtful that, with a reused salt, a rainbow table of reasonably short human-readable passwords would be that expensive to construct.

That said, I am by no means an expert in this stuff.

Reply:
Hello there! I'm Akash, and I've been thinking about contributing to the Taskwarrior repository for a while. After reading this discussion, I have an idea to share:

I looked into using a robust salt for all passwords. When I say "robust" I mean a lengthy combination of letters, numbers, and symbols. Doing some quick math, with a 10-character salt containing uppercase, lowercase, numbers, and symbols, we end up with about 630 quadrillion possible combinations.

Assuming a state actor has a table of all possible 10-character passwords, that's another 630 quadrillion. Now, if they want to calculate all possible hashes for all possible salts, it would require computing over 450 octillion hashes, which, I assume, would take an eternity.

These calculations are for a single hashing iteration. Since we'll be doing 600,000 iterations, using a strong salt seems like a solid choice.

I also thought of another idea. What if we create a simple hashing function that returns a number between 0 and the length of the password? This number would be the index where the salt is inserted. Here's an example:

salt: a1B4c9
password: thisisawesome@123
length of password: 17
our hashing function h1(k) -> (maybe) [Polynomial Accumulation](https://rentry.co/6nfgs) of password % length of password
let h1("thisisawesome@123") = 3
then salted password is thia1B4c9sisawesome@123 (salt inserted at the third index)
the salted password will be further hashed by some algorithm like SHA256

The only downside I could think for this method is to create a consistent hashing function. I am not sure if my idea is already a thing though.

I would love your opinions @djmitche @ryneeverett !!

Reply:
OK, I think we're roughly agreeing here -- the way we are using PBKDF doesn't benefit from salt.  Since a value is required for the algorithm, we can just use a fixed value with no harm.

The manpage recomments using `pwgen` to generate secrets, so I think those will be suitably random. We can bump to 600,000 iterations in a subsequent PR.

Probably the term I should have used was "advanced persistent threat"  - an adversary willing to put a _lot_ of resources into a specific target, instead of fishing for easy targets. Various spy agencies fit the bill, and I suppose a cloud provider could, too -- although I don't see AWS deciding it's worth the reputational and financial cost to generate Taskwarrior-specific rainbow tables.

@akashShanmugraj thanks for the analysis! Even with just a password, we can ensure a lot of iterations are required to find the one used to generate the key used for the AEAD encryption.

I think we should avoid trying to invent our own algorithms -- that often introduces subtle flaws. We should stick to existing, vetted algorithms instead.  In fact, that's why we picked PBKDF!

Reply:
Thanks for jumping in @akashShanmugraj.

The point I was trying to make was that a 10-character salt and a 10-character password have the same number of possible combinations as a 20-character password.

A piece of context you might not have is that the only idea we have for sharing the salt amongst all clients is by including it in the configuration file, which is the same mechanism by which the password is distributed. In this scenario we agree that having users distribute a unique salt doesn't add any value.

However, it now occurs to me that another option for cloud sync would be to generate and store a unique salt in the clear in cloud storage under a "salt" key and add an endpoint to let replicas retrieve it. This would prevent construction of a single "taskwarrior" rainbow table to attack all users and seems aligned with the purpose and requirements of a salt. What do you think @djmitche?

> Probably the term I should have used was "advanced persistent threat" - an adversary willing to put a lot of resources into a specific target, instead of fishing for easy targets.

In this context, an APT might construct a rainbow table for a single user whereas a more opportunistic attacker might build a rainbow table for all taskwarrior users if a single salt is used. I believe salts are useless against an APT.

> Various spy agencies fit the bill, and I suppose a cloud provider could, too -- although I don't see AWS deciding it's worth the reputational and financial cost to generate Taskwarrior-specific rainbow tables.

Just to be clear, I wasn't singling out cloud providers because they have large compute resources, I was singling out cloud providers because we're literally giving them our data when using cloud sync. I think the biggest threat is probably that they will accidentally allow an "AI" to crack our encryption.